ndcube/mixins/ndcubeio.py

```diff
@@ -0,0 +1,257 @@
+            wcs_header = self.__wcs.to_header(relax=True)
+            header.extend(wcs_header, useblanks=False, update=True)
+
+        # Create a FITS header for storing missing_axis
+        # MISSING0 : 1 if axis 1 is missing
+        # other keywords are skipped, if the axis is present
+        if self.__missing_axis:
+            header0 = fits.Header()
+            for index, axis in enumerate(self.__missing_axis):
+                if axis:
+                    header0[MISSING.format(index)] = 1
+                else:
+                    continue
+
+            for k, v in header0.items():
+                _insert_in_metadata_fits_safe(header0, k, v)
+
+            header.extend(header0, useblanks=False, update=True)
+
+
+        # PrimaryHDU list contains only meta, missing_axis and wcs as headers, no data
+        hdus = [fits.PrimaryHDU(data=None, header=header)]
+
+        #------------------------------HDU0----------------------------------
+
+        #------------------------------HDU1----------------------------------
+        # Header for unit
+        if self.__unit:
+            header_unit = fits.Header()
+            if self.__unit is not u.dimensionless_unscaled:
+                header_unit['bunit'] = self.__unit.to_string()
+            hdus.append(fits.ImageHDU(self.__data, header_unit, name='UNIT'))
+
+        #------------------------------HDU1----------------------------------
+
+        #------------------------------HDU2----------------------------------
+        # Store the uncertainty
+        if self.__uncertainty is not None:
+            
+            # Set the initial header, and the type of uncertainty
+            hdr_uncertainty = fits.Header()
+            hdr_uncertainty['UTYPE'] = self.__uncertainty.uncertainty_type
+
+            # Set the data of the uncertainty and header
+            hduUncert = fits.ImageHDU(self.__uncertainty.array, hdr_uncertainty, name='UTYPE')
+            hdus.append(hduUncert)
+        #----------------------------HDU2--------------------------------
+
+        #----------------------------HDU3------------------------------
+        # Store the mask
+        if self.__mask is not None:
+            # Always assuming that the mask is a np.ndarray
+            # by checking that it has a shape
+            if not hasattr(self.__mask, 'shape'):
+                raise ValueError('only a numpy.ndarray mask can be saved.')
+            hduMask = fits.ImageHDU(self.__mask.astype(np.uint8), name=hdu_mask)
+            hdus.append(hduMask)
+
+        #----------------------------HDU3-------------------------------
+        # Store the extra_coords
+        if self.__extra_coords is not None:
+
+            # Set up the data
+            flattened_list_of_dict = flatten(self.__extra_coords)
+
+            # We convert the list of dictionary to pandas dataframe
+            # and then to numpy.ndarray
+            dframe = pd.DataFrame(flattened_list_of_dict)
```

Comment:
Can this be done without introducing a dependency on pandas?

Also, can you quickly confirm how this stored the values?  Is it a table where the value/array of an extra coord is stored in a single cell in the table?  Remember all extra coords are not the same length/shape depending on the axis or axes to which they correspond.

Reply:
There should be another column in this table giving the unit of each extra coord.  If the extra coord value is not stored in a ```astropy.unit.Quantity```, the unit can be ```None```.

Reply:
Since I am flattening the dictionary, in return I am getting a dictionary of all key/value pairs. So I am storing them as a list of headers, with each entry as the key and the value of the header is the value of the key.
> Remember all extra coords are not the same length/shape depending on the axis or axes to which they correspond.

Regarding this, can you elaborate, as I did not understand? :sweat_smile:

Reply:
Sure. Say you have a 3D NDCube with shape (3, 4, 5). Now say you have an extra coords, ECA, along the 1st axis. There is a value for each pixel along the 1st axis so the time extra coord is of shape (3,). Now say you have another extra coord along the 2nd axis, ECB. It will have a shape of (4,), as that’s the length of the 2nd axis. So ECA and ECB don’t fit into a table unless there’s an additional blank value somewhere for ECA.

However it can be more complicated than that. The two above extra coords only apply to one axis each. Say we have a 3rd extra coord, ECC, whose value depends on both the 2nd and 3rd axes. In this case ECC will have a shape of (4, 5), since those are the lengths of the2nd and 3rd axes. If this doesn’t make sense at first, a practical example of coords like this are latitude and longitude. The longitude in a given pixel column (x) depends also on the pixel row (y). In this case ECC can not be easily stored in a table if each pixel value needs its own cell in the table. The only way is if one cell contains the entire 1- or 2-D array of extra coord values.

Reply:
Can we have an extra_coords of (m,n,o)(3-dimensions), or an extra_coords of more than 3 dimensions?

I guess we should save each of the extra_coords in TableHDU, as each of the extra_coords are unique, and it would be better to store them based on their shape.

> Can we have an extra_coords of (m,n,o)(3-dimensions), or an extra_coords of more than 3 dimensions?

If the above configuration is possible, we can reduce them to 2 dimensional table entries.

Reply:
Storing the extra_coords, ECA, ECB, ECC, and so on, would certainly be easy to manage.

Reply:
@yashrsharma44 An extra coord can have any number of dimensions up to the dimensionality of the data.  So if the data is 3D an extra coord can be 1, 2 or 3 D.  If the data is 5D, an extra coord can be 1, 2, 3, 4 or 5 D etc.

Reply:
Also, can I follow up on the question of whether we really need to introduce a dependency on pandas.

Reply:
Umm, we can drop the pandas dependency, I would code up to get the flattened values

Reply:
I think I spoke to this on matrix a little. I think the best thing to do here is to a) use `astropy.table.QTable` instead of pandas. and b) to make a table of extra coords where the arrays in the extra coords are in a single cell of the FITS table.

i.e. you would have a table that looks like this:
| name     | unit     | data           |
| -------- | -------- | --------       |
| em.wl    | nm       | [[1,2], [3,4]] |
| time     | sec      | [100, 200, 30

Reply:
Great! Thanks for the suggestions, I would implement them right away!